File: backend/internal/models/student.go
package models

type CreateStudentRequest struct {
	// Asgardeo account fields
	Email       string `json:"email" binding:"required,email"`
	GivenName   string `json:"given_name" binding:"required"`
	FamilyName  string `json:"family_name" binding:"required"`
	PhoneNumber string `json:"phone_number"`
	Password    string `json:"password" binding:"required,min=8"`

	// Student profile fields
	IndexNumber    string `json:"index_number" binding:"required"`
	Address        string `json:"address"`
	WhatsApp       string `json:"whatsapp"`
	SpecialRemarks string `json:"special_remarks"`
	// Gender is optional; when set it must be one of the DB's allowed values.
	Gender string `json:"gender" binding:"omitempty,oneof=male female"`
}

type UpdateStudentRequest struct {
	GivenName      string `json:"given_name" binding:"required"`
	FamilyName     string `json:"family_name" binding:"required"`
	PhoneNumber    string `json:"phone_number"`
	Address        string `json:"address"`
	WhatsApp       string `json:"whatsapp"`
	SpecialRemarks string `json:"special_remarks"`
	Gender         string `json:"gender" binding:"omitempty,oneof=male female"`
}

type StudentResponse struct {
	ID             string `json:"id"`
	UserID         string `json:"user_id"`
	FullName       string `json:"full_name"`
	IndexNumber    string `json:"index_number"`
	Address        string `json:"address"`
	Phone          string `json:"phone"`
	WhatsApp       string `json:"whatsapp"`
	SpecialRemarks string `json:"special_remarks"`
	Gender         string `json:"gender"`
	CreatedAt      string `json:"created_at"`
}

type StudentWithClassResponse struct {
	StudentResponse
	ClassName    string `json:"class_name"`
	GradeName    string `json:"grade_name"`
	AcademicYear string `json:"academic_year"`
}

File: frontend/src/services/student.ts
import api from "../lib/api";

export interface Student {
  id: string;
  user_id: string | null;
  full_name: string;
  index_number: string;
  address: string | null;
  phone: string | null;
  whatsapp: string | null;
  special_remarks: string | null;
  gender: string | null;
  created_at: string | null;
  updated_at: string | null;
  class_name: string | null;
  grade_name: string | null;
  house_id: string | null;
  house_name: string | null;
}

export interface StudentWithClass extends Student {
  class_name: string | null;
  grade_name: string | null;
  academic_year: string | null;
}

export interface CreateStudentRequest {
  email: string;
  given_name: string;
  family_name: string;
  phone_number?: string;
  password: string;
  index_number: string;
  address?: string;
  whatsapp?: string;
  special_remarks?: string;
  gender?: "male" | "female";
}

export interface UpdateStudentRequest {
  given_name: string;
  family_name: string;
  phone_number?: string;
  address?: string;
  whatsapp?: string;
  special_remarks?: string;
  gender?: "male" | "female";
}

export const studentApi = {
  list: () => api.get<Student[]>("/students").then((r) => r.data),

  get: (id: string) => api.get<Student>(`/students/${id}`).then((r) => r.data),

  getWithClass: (id: string) =>
    api.get<StudentWithClass>(`/students/${id}/class`).then((r) => r.data),

  listByClass: (classId: string) =>
    api.get<Student[]>(`/classes/${classId}/students`).then((r) => r.data),

  create: (data: CreateStudentRequest) =>
    api.post<Student>("/students", data).then((r) => r.data),

  update: (id: string, data: UpdateStudentRequest) =>
    api.put<Student>(`/students/${id}`, data).then((r) => r.data),

  updateHouse: (id: string, houseId: string) =>
    api
      .put<Student>(`/students/${id}/house`, { house_id: houseId })
      .then((r) => r.data),

  remove: (id: string) =>
    api.delete(`/students/${id}`).then((r) => r.data),
};
